app/assets/javascripts/components/new_restaurant.jsx

```diff
@@ -0,0 +1,171 @@
+
+class NewRestaurant extends React.Component {
+
+    constructor() {
+        super()
+        this.state = {
+            formError: null,
+            edit: false
+        }
+    }
+
+    componentWillMount() {
+        Formsy.addValidationRule('isValidAddress', (values, value) => {
+            var geocoder = new google.maps.Geocoder();
+            geocoder.geocode({'address': value}, function(results, status) {
+                debugger;
+                return (status === google.maps.GeocoderStatus.OK)
+            });
+        });
+        this.state = {
+            formError: null,
+            edit: this.props.hasOwnProperty("params") && this.props.params.hasOwnProperty("id")
+        }
+    }
+
+
+    onValidSubmit(model) {
+        console.log("valid")
+
+        new google.maps.Geocoder().geocode({'address': model.address}, function(results, status) {
+            if (status === google.maps.GeocoderStatus.OK) {
+                model.lat = results[0].geometry.location.lat();
+                model.lang = results[0].geometry.location.lng();
+                if (this.state.edit) {
+                    Api.updateRestaurant(this.props.params.id, model.name, model.food_type,
+                        model.speed, model.ten_bis ? model.ten_bis : false, model.delivery ? model.delivery : false,
+                        model.rank, model.link, model.description, model.address, model.lat, model.lang)
+                } else {
+                    Api.createRestaurant(model.name, model.food_type,
+                        model.speed, model.ten_bis ? model.ten_bis : false, model.delivery ? model.delivery : false,
+                        model.rank, model.link, model.description, model.address, model.lat, model.lang)
+                }
+                this.context.router.transitionTo("start")
+            }else {
+                window.alert("Invalid address, google does not know" + model.address + " " + status);
+            }
+        }.bind(this));
+    }   
+
+    onInvalidSubmit(message) {
+        console.log("invalid")
+        this.setState ({
+           formError: message
+       })
+    }
+
+    onChange() {
+
+    }
+
+    render() {
+
+        let restaurant =  this.state.edit ? this.props.restaurants.find( r => r.id == this.props.params.id): {};
```

Comment:
For nice but, unfortunately, not fully-supported functions like 'find' we can use something like [lodash](https://lodash.com/) to be sure than this will work cross-browser.